src/optimizer_interface.jl

```diff
@@ -114,15 +112,9 @@ function optimize!(model::Model,
     end
 
     if optimizer_factory !== nothing
-        if mode(model) == DIRECT
-            error("An optimizer factory cannot be provided at the `optimize` call in DIRECT mode.")
-        end
-        if MOIU.state(backend(model)) != MOIU.NO_OPTIMIZER
-            error("An optimizer factory cannot both be provided in the `Model` constructor and at the `optimize` call.")
```

Comment:
We could also just remove this condition and `optimize!(model, with_optimizer(...))` would be a shortcut for `set_optimizer(with_optimizer(...)); optimizer!(model)`. What would be wrong with that ?

Reply:
This won't work when we remove `with_optimizer`, which I would like to do after this PR. I can toss this PR and go straight to removing `with_optimizer` if that makes more sense.

Reply:
> This won't work when we remove `with_optimizer`

You mean when we replace `Model(with_optimizer(GLPK.Optimizer))` by `Model(GLPK.Optimizer)` ? Why wouldn't `optimize!(model, GLPK.Optimizer)` work ?

Reply:
The non-obvious part is how to deal with parameters. We could take a dictionary and automate the call to `set_parameter` for each key/value in the dictionary. 

However, I think there's a bigger issue this API. Keeping the discussion to `with_optimizer`, consider:
```julia
optimize!(model, with_optimizer(GLPK.Optimizer, some_parameter=some_value))
...
optimize!(model)
```
It's not really clear that the solver or parameter choice in the first call to `optimize!` would apply to the second call. Indeed, this is inconsistent with cvxpy and Convex.jl where I believe that the second call would fall back to a default solver with default parameters instead of the last solver used. Hiding the fact that the JuMP model is stateful with respect to the optimizer will cause more confusion like https://github.com/JuliaOpt/JuMP.jl/issues/2074.

Reply:
The only way to make it non-confusing is to keep a flag in the JuMP Model that the optimizer should be discarded before the next `optimize!` call (we need to keep it until then in case the user query results). I agree that we don't really want to go in that direction.